### PR TITLE
[AMD][HIP] NSA: bf16 passthrough from RMSNorm to eliminate FP8 dequantization

### PR DESCRIPTION
**Pull request by Jacob0226**
## Motivation

When running GLM-5-FP8 on MI355X with FP8 quantization, the NSA indexer's head gate path (`weights_proj`) requires bf16 input. However, the upstream `fused_rms_fp8_group_quant` only outputs FP8 tensors, discarding the bf16 intermediate from RMSNorm. This forces a redundant FP8-to-bf16 dequantization (4 PyTorch eager kernels, ~18 us per layer) that reconstructs a value already computed inside RMSNorm.

## Modifications

- **`communicator.py`**: When NSA is active on gfx95 with FP8 quantization, set `output_unquantized_inp1=True` in `fused_rms_fp8_group_quant` to preserve the bf16 output at near-zero cost. Pack the result as a 3-tuple `(x_fp8, x_scale, x_bf16)`.
- **`nsa_indexer.py`**: In `_weights_proj_bf16_in_fp32_out`, extract `x[2]` (bf16) directly from the 3-tuple on HIP, completely bypassing dequantization. Skip the dequant block in `forward_cuda` for 3-tuple inputs. Also enable `torch.compile` on HIP for `_get_logits_head_gate` and `_project_and_scale_head_gates` (aligned with PR #22232).

The 3-tuple flows transparently through the existing tuple-handling paths: `Fp8LinearMethod.apply` uses only `x[0]` and `x[1]` for FP8 GEMMs (e.g., `self.wk`), so `x[2]` is naturally ignored by quantized layers.

This change affects the AMD HIP path only.
image

## Accuracy Tests
GSM8K accuracy:
- GLM-5-FP8 on MI355X (TP8): 0.945
- GLM-5-FP8 on B200 (TP8): 0.949

## Speed Tests and Profiling

GLM-5-FP8 server cmd on MI355:
```
export SGLANG_ROCM_FUSED_DECODE_MLA=0
export ROCM_QUICK_REDUCE_QUANTIZATION=INT4
export SAFETENSORS_FAST_GPU=1
python3 -m sglang.launch_server \
  --model-path GLM-5-FP8 \
  --tp 8 \
  --port 9000 --trust-remote-code \
  --tool-call-parser glm47 --reasoning-parser glm45 \
  --mem-fraction-static 0.85 \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}' \
  --nsa-prefill-backend tilelang --nsa-decode-backend tilelang --disable-radix-cache \
  --kv-cache-dtype fp8_e4m3
```

Benchmark on MI355X TP8, concurrency 4/8/16/32/64 averaged (baseline: sglang PR #22232 + aiter PR#2575):

- **ISL/OSL 1k/1k**: Throughput **+3.4%**, TPOT **+3.8%**
- **ISL/OSL 8k/1k**: Throughput **+2.8%**, TPOT **+1.0%**

Per-layer profiling:
- FP8 dequantization: ~18 us → **0 us** (4 kernels eliminated)

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by HaiShaw — CHANGES_REQUESTED** [review]
Please align the producer use `_use_aiter and _is_gfx95_supported`, and consumer use `_is_hip`.  After #22422 you can use `_use_aiter_gfx95`

**Comment by Jacob0226** [comment]
> Please align the producer use `_use_aiter and _is_gfx95_supported`, and consumer use `_is_hip`. After #22422 you can use `_use_aiter_gfx95`

Done. Updated the consumer guard in nsa_indexer.py to use _use_aiter and _is_gfx95_supported, aligned with the producer side in communicator.py.

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
@amd-bot ci-status
